lib/AbstractHandle/Utils/Handler.py

```diff
@@ -93,20 +94,23 @@ def _get_token_roles(self, token):
             return customroles
 
     def _is_admin_user(self, token):
-        fetched_token_info = self.token_cache.get(token)
+        with self._cache_lock:
+            fetched_token_info = self._token_cache.get(token)
 
         if fetched_token_info:
             customroles = fetched_token_info.get('customroles')
         else:
             customroles = self._get_token_roles(token)
-            self.token_cache[token] = {'customroles': customroles}
+            with self._cache_lock:
+                self._token_cache[token] = {'customroles': customroles}
```

Comment:
Does inserting also need a thread lock?

Reply:
Yes, that's where the race condition is happening - expire() is called on insert